presentation/frontend/src/test/StreamedResponseGenerator.test.ts

```diff
@@ -0,0 +1,225 @@
+import { StreamedResponseGenerator, StreamedResponseGeneratorProps } from "../presentation-frontend/StreamedResponseGenerator";
+import { expect } from "chai";
+import sinon from "sinon";
+import { PagedResponse, PageOptions } from "@itwin/presentation-common";
+import { ResolvablePromise } from "@itwin/presentation-common/lib/cjs/test";
+
+describe("StreamedResponseGenerator", () => {
+  it("should run requests concurrently", async () => {
+    const total = 10;
+    const firstBatchPromise = new ResolvablePromise<PagedResponse<number>>();
+    const restBatchesPromise = new ResolvablePromise<PagedResponse<number>>();
+    const fakeGetBatch = sinon.fake(async (_, idx: number) => {
+      const result = idx ? restBatchesPromise : firstBatchPromise;
+      return result as unknown as Promise<PagedResponse<number>>;
+    });
+
+    const generator = new StreamedResponseGenerator({ getBatch: fakeGetBatch });
+    const getItemsPromise = generator.createItemsResponse();
+    expect(fakeGetBatch).to.be.calledOnce;
+    expect(fakeGetBatch).to.be.calledWith({ start: 0, size: 0 }, 0);
+
+    await firstBatchPromise.resolve({ total, items: [1, 2] });
+    const expectedCallCount = total / 2;
+    expect(fakeGetBatch.callCount).to.eq(expectedCallCount);
+    const expectedCalls = [...new Array(expectedCallCount - 1).keys()].map((i) => [{ start: (i + 1) * 2, size: 2 }, i + 1]);
+    const actualCalls = fakeGetBatch
+      .getCalls()
+      .slice(1)
+      .map((x) => x.args);
+    expect(actualCalls).to.deep.eq(expectedCalls);
+
+    await restBatchesPromise.resolve({ total, items: [3, 4] });
+    const expectedResult = { total, items: [1, 2].concat(...[...new Array(expectedCallCount - 1).keys()].map(() => [3, 4])) };
+    await expect(getItemsPromise).to.eventually.deep.eq(expectedResult);
+  });
+
+  it("should return values in a correct order", async () => {
+    const total = 4;
+    const promiseResolutionOrderings = [
+      [0, 1, 2],
+      [0, 2, 1],
+      [1, 0, 2],
+      [1, 2, 0],
+      [2, 0, 1],
+      [2, 1, 0],
+    ];
+
+    for (const ordering of promiseResolutionOrderings) {
+      const fakePromises = [...new Array(total - 1).keys()].map(() => new ResolvablePromise());
+      const generator = new StreamedResponseGenerator({
+        getBatch: async (_, idx) => {
+          if (idx === 0) {
+            return { total, items: [0] };
+          }
+
+          return fakePromises[idx - 1] as unknown as Promise<PagedResponse<unknown>>;
+        },
+      });
+
+      const itemsPromise = generator.createItemsResponse();
+      for (const idx of ordering) {
+        await fakePromises[idx].resolve({ total, items: [idx + 1] });
+      }
+
+      const { items } = await itemsPromise;
+      expect(items).to.deep.eq([...new Array(total).keys()]);
+    }
+  });
```

Comment:
I don't think we need to check every possible combination of promise resolution order. For the same of simplicity, I'd:
- Rename the test to `returns values in correct order when requests resolve when requests resolve in different order than being made`.
- Use 2 resolvable promises for 2nd and 3rd pages, resolve 3rd first and then 2nd.
- Check the result.

Reply:
That's not enough. Apparently the edge case when there is a difference of more than 1 between batch indices is left uncovered (`batch.idx - 1 !== lastEmitted`). I still have to use 3 promises, but I'll only leave 2 orderings, which will cover all the edge cases.